### PR TITLE
Migrated `pkg/controller/endpoint` to contextual logging

### PR DESCRIPTION
**Pull request by my-git9**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Ref https://github.com/kubernetes/enhancements/issues/3077

#### Does this PR introduce a user-facing change?

```release-note
Migrate `pkg/controller/endpoint` to contextual logging
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

### CODE REVIEW

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -322,37 +323,39 @@ func (e *Controller) processNextWorkItem(ctx context.Context) bool {
...
 
 	err := e.syncService(ctx, eKey.(string))
-	e.handleErr(err, eKey)
+	e.handleErr(ctx, err, eKey)
 
 	return true
 }
 
-func (e *Controller) handleErr(err error, key interface{}) {
+func (e *Controller) handleErr(ctx context.Context, err error, key interface{}) {
```
Comment (pohly) [inline pkg/controller/endpoint/endpoints_controller.go:331]: I think this is one of those cases where passing `logger klog.Logger` is a better fit for the function.

File: pkg/controller/endpoint/endpoints_controller.go [hunk 2]
```diff
@@ -322,37 +323,39 @@ func (e *Controller) processNextWorkItem(ctx context.Context) bool {
...
 
 	if e.queue.NumRequeues(key) < maxRetries {
-		klog.V(2).InfoS("Error syncing endpoints, retrying", "service", klog.KRef(ns, name), "err", err)
+		logger.V(2).Info("Error syncing endpoints, retrying", "service", klog.KRef(ns, name), "err", err)
 		e.queue.AddRateLimited(key)
 		return
 	}
 
-	klog.Warningf("Dropping service %q out of the queue: %v", key, err)
+	logger.Info("Dropping service out of the queue", "service", key, "queue", err)
```
Comment (pohly) [inline pkg/controller/endpoint/endpoints_controller.go:349]: Let's use "service" for an object and "key" for a work item key string.

File: pkg/controller/endpoint/endpoints_controller.go [hunk 3]
```diff
@@ -583,11 +586,12 @@ func (e *Controller) checkLeftoverEndpoints() {
 // The addresses are added to the corresponding field, ready or not ready, depending
 // on the pod status and the Service PublishNotReadyAddresses field value.
 // The pod passed to this function must have already been filtered through ShouldPodBeInEndpoints.
-func addEndpointSubset(subsets []v1.EndpointSubset, pod *v1.Pod, epa v1.EndpointAddress,
+func addEndpointSubset(ctx context.Context, subsets []v1.EndpointSubset, pod *v1.Pod, epa v1.EndpointAddress,
```
Comment (pohly) [inline pkg/controller/endpoint/endpoints_controller.go:589]: Let's pass `logger klog.Logger` here.

**Review by pohly — CHANGES_REQUESTED** [review]

**Review by pohly — CHANGES_REQUESTED** [review]
Almost done :smile: 

Can you also remove the exception from https://github.com/kubernetes/kubernetes/blob/5f23f835879b514e9bf16809059b38957d20a486/hack/logcheck.conf#L45

**Comment by my-git9** [comment]
> Almost done 😄
> 
> Can you also remove the exception from
> 
> https://github.com/kubernetes/kubernetes/blob/5f23f835879b514e9bf16809059b38957d20a486/hack/logcheck.conf#L45

Updated, thanks for your review

**Review by pohly — APPROVED** [review]
/lgtm
/assign @soltysh 

For approval.

File: pkg/controller/endpoint/endpoints_controller.go [hunk 4]
```diff
@@ -321,38 +322,40 @@ func (e *Controller) processNextWorkItem(ctx context.Context) bool {
...
 	e.queue.Forget(key)
 	utilruntime.HandleError(err)
 }
 
 func (e *Controller) syncService(ctx context.Context, key string) error {
 	startTime := time.Now()
+	logger := klog.FromContext(ctx)
 	defer func() {
-		klog.V(4).Infof("Finished syncing service %q endpoints. (%v)", key, time.Since(startTime))
+		logger.V(4).Info("Finished syncing service endpoints", "service", key, "startTime", time.Since(startTime))
```
Comment (pohly) [inline pkg/controller/endpoint/endpoints_controller.go:358]: /lgtm cancel

I missed this initially, but then saw it in another PR: `service` is used inconsistently. Sometimes it's the work queue key (a string), sometimes the object (from `klog.KObj`).

Let's use `key` for the string and `service` for the object. 

This also applies in some other places, please search/replace.

Comment (my-git9) [reply pkg/controller/endpoint/endpoints_controller.go]: Updated, thanks

**Comment by mengjiao-liu** [comment]
ping @soltysh

**Review by pohly — APPROVED** [review]
/lgtm

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**